ts/src/profiler.ts

```diff
@@ -13,47 +13,292 @@
+    return this.pollProfilerService();
+  }
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   */
+  async pollProfilerService(): Promise<void> {
+    const startCreateMillis = Date.now();
+    const prof = await this.createProfile();
+    await this.profileAndUpload(prof);
+    const endCreateMillis = Date.now();
+
+    // Schedule the next profile.
+    setImmediate(this.pollProfilerService.bind(this)).unref();
+  }
+
+  /**
+   * Talks to Stackdriver Profiler server, which hangs until server indicates
+   * job should be profiled.
+   *
+   * If any problem is encountered, the problem will be logged and
+   * createProfile() will be retried.
+   *
+   * TODO: implement backoff and retry when error encountered. createProfile()
+   * should be retried at time response indicates this request should be retried
+   * or with exponential backoff (up to one hour) if the response does not
+   * indicate when to retry this request. Once this is implemented, an error
+   * will be thrown only if the error indicates one definitely should not
+   * retry createProfile.
+   *
+   * Public to allow for testing.
+   */
+  async createProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: {
+        projectId: this.config.projectId,
+        target: this.config.serviceContext.service,
+        labels: {zone: this.config.zone, instance: this.config.instance},
+      },
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    let requestError: Error|undefined = undefined;
+    let retryRequest = true;
+    try {
+      const [body, response] = await this.request(options);
+      if (isErrorResponseCode(response.statusCode)) {
+        retryRequest = isRetriableResponseCode(response.statusCode);
+        requestError =
+            new Error('Error creating profile: ' + response.statusMessage);
+        this.logger.debug(requestError);
+      } else {
+        return body;
+      }
+    } catch (err) {
+      retryRequest =
+          isRetriableError(err) || isRetriableResponseCode(err.statusCode);
+      requestError = new Error('Error creating profile: ' + err.toString());
+      this.logger.debug(requestError);
+    }
+    if (retryRequest) {
+      // TODO: check response to see if response specifies a backoff.
+      // TODO: implement exponential backoff.
+      await delay(this.config.backoffMillis);
+      return this.createProfile();
+    }
+    throw requestError;
+  }
+
+  /**
+   * Collects a profile of the type specified by the profileType field of prof.
+   * If any problem is encountered, like a problem collecting or uploading the
+   * profile, an error will be logged at the debug level, but otherwise ignored.
+   *
+   * Public to allow for testing.
+   *
+   * @param prof
```

Comment:
What's the point of this "@param prof" markup I am curious? It does not explain anything, does not document the type, and I can clearly see that the function has one parameter named `prof` from the source.

Reply:
Done. I removed this.